system/src/system_power_manager.cpp

```diff
@@ -552,52 +526,94 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
+            handleStateChange(g_batteryState, batteryState);
+          }
         } else {
-          PMIC power;
-          power.setRechargeThreshold(PMIC_FAULT_RECHARGE_THRESHOLD);
-          possibleFaultCounter_ = 0;
-          faultSecondaryCounter_ = 1;
-          possibleFaultTimestamp_ = millis();
+          vcellDebounceCount_ = 0;
         }
       }
     }
-  }
-  return to;
-}
-
-void PowerManager::handlePossibleFaultLoop() {
-  if (faultSecondaryCounter_ == 1 && (millis() - possibleFaultTimestamp_ > DEFAULT_FAULT_WINDOW)) {
-      PMIC power;
-      power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD);
-      faultSecondaryCounter_ = 0;
-      faultSuppressed_ = millis();
+  } else {
+    // Charging is enabled
+    battery_state_t batteryState;
+    uint8_t status;
+    PMIC power(true);
+    FuelGauge fuel(true);
+    if (chargedTimeStamp_ != 0 && millis() - chargedTimeStamp_ >= BATTERY_CHARED_POSTPONE_PERIOD) {
+      chargedTimeStamp_ = 0;
+      status = power.getSystemStatus();
+      // When we get here, a charged event is raised previously followed by disabling charging,
+      // so the VBAT is not floating and the VSYS_STAT can be used to deduce the battery state.
+      // If battery is actually charged, then the battery voltage should above VSYS_MIN (3.5v)
+      // The VSYS_STAT bit should be cleared, otherwise, the battery is disconnected
+      if (status & 0x01) {
+        batteryState = BATTERY_STATE_DISCONNECTED;
+      } else {
+        // The battery may be attached during the debounce period.
+        // We need to re-check the status if device is under charging actually.
+        power.enableCharging();
+        power.enableSafetyTimer();
+        const uint8_t chrg_stat = (power.getSystemStatus() >> 4) & 0b11;
+        if (chrg_stat == 0b01 || chrg_stat == 0b10) {
+          batteryState = BATTERY_STATE_CHARGING;
+        } else {
+          batteryState = BATTERY_STATE_CHARGED;
+        }
+      }
+      handleStateChange(g_batteryState, batteryState);
+    } else if (g_batteryState == BATTERY_STATE_DISCONNECTED && millis() - disconnectedTimeStamp_ >= BATTERY_STATE_CHECK_PERIOD) {
+      disconnectedTimeStamp_ = millis();
+      if (chargingDebounceCount_ == 0) {
+        power.enableCharging();
+        power.enableSafetyTimer();
+      }
+      // If battery is of low voltage, it won't trigger INT, instead we need to poll the charging state
+      const uint8_t chrg_stat = (power.getSystemStatus() >> 4) & 0b11;
+      if (chrg_stat == 0b01 || chrg_stat == 0b10) {
+        chargingDebounceCount_++;
+        if (chargingDebounceCount_ >= BATTERY_CHARGING_DEBOUNCE_COUNT) {
+          batteryState = BATTERY_STATE_CHARGING;
+          handleStateChange(g_batteryState, batteryState);
+        }
+      } else if (chrg_stat == 0b00) {
+        chargingDebounceCount_ = 0;
+      } else {
+        // if battery is not attached, the charged event will probably trigger INT and followed by disabling charging
```

Comment:
> and followed by disabling charging

We will also get here if the re-attached battery is above recharge threshold. We should clarify the comment that this state change will correctly be handled by `handleUpdate()`.